[AIRFLOW-1688] Automatically add load.time_partitioning to bigquery_hook when table name includes $

albertocalderari opened the pull request:
Dear Airflow maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!
Allow automatic creation of a new partitioned tables when the '$' symbol is in the table name adding the option:

load.timePartitioning: {type: 'DAY'}

i.e. my_dataset.my_table$20170101 would be automatically loaded as a new partitioned table without the need of creating it manually.

Dear Airflow maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!


### JIRA
- [x] My PR addresses the following [Airflow JIRA](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "[AIRFLOW-1688] Automatically add load.time_partitioning to bigquery_hook when table name includes $"
    - https://issues.apache.org/jira/browse/AIRFLOW-1688


### Description
- [x] Here are some details about my PR, including screenshots of any UI changes:
The gcs_to_bq operator throws and exception when trying to auto-create a new date partitioned table. 
To allow the table creation the load configuration needs the api option:
load.timePartitioning:
{type: 'DAY'}
I will add a fix to identify date partitioned table from the presence of a $ in the table name and add the option.
I have also added "time_partitioning" option to the load, which is a dictionary that mirrors the 'timePartitioning' config in google's api. Using the time partitioning option we can add on top of partition type also a specific field to partition by and the partition expiration time as per example below:

{'type': 'DAY', 'field': 'test_field', 'expirationMs': 1000}

### Tests
- [x] My PR adds the following unit tests:
added unitest case TestTimePartitioningInRunJob in tests.contrib.hooks.test_bigquery_hook to verify that the normal job execution won't fail because of the changes, '$' in the table name adds automatically the partition configuration to the load, the extra time partitioning options are returned as expected, field based partitioning and specific partition name can't be specified at the same time

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

Commit:
[AIRFLOW-1688] Add extra option to gcs to bq operator

Commit:
[AIRFLOW-1688] Add partition handling on run_load in the hook

Add unit test for partition options in big query hook

jgao54 commented on airflow/contrib/hooks/bigquery_hook.py:
```diff
@@ -440,6 +443,10 @@ def run_load(self,
+        :param time_partitioning: configure optional time partitioning fields i.e. partition by field, type and
+            expiration as per API specifications. note that 'field' is not available in concurrency with
+            dataset.table$partition.
+        :type src_fmt_configs: dict
```
nit: time_partitioning

jgao54 commented on airflow/contrib/hooks/bigquery_hook.py:
```diff
@@ -487,6 +494,19 @@ def run_load(self,
+            if not configuration.get('load', {}).get('timePartitioning'):
+                configuration['load']['timePartitioning'] = {}
+            configuration['load']['timePartitioning'].update(time_partitioning)
+
```
Hmmm my thought is we should keep `timePartitioning` optional even when table is partitioned (unless there is a good reason to always use timePartitioning for partitioned table). This could simplify the code as well:

```
if time_partitioning:
    if '$' in destination_project_dataset_table and time_partitioning.get('field'):
        raise AirflowException("Cannot specify field partition and partition name (dataset.table$partition) at the same time")
    configuration['load']['timePartitioning'].update(time_partitioning)
```

jgao54 replied:
actually scratch what I said, automatic creation of a new partitioned tables makes sense! :D

jgao54 reviewed (approved):
just a small typo, but otherwise lgtm!

Commit:
[AIRFLOW-1688] Fix the typo and the conflicts

dsdinter committed:
[AIRFLOW-1954] Add DataFlowTemplateOperator

Closes #2909 from dsdinter/dataflow_template

walternewzoo committed:
added Newzoo to airflow organizations

Closes #2917 from
walternewzoo/airfloworganizations-newzoo

Commit:
[AIRFLOW-1969] Always use HTTPS URIs for Google OAuth2

Closes #2900 from intellectronica/google-auth-
force-scheme

edgarRd committed:
[AIRFLOW-1971] Propagate hive config on impersonation

Currently, if hive specific settings are defined
in the configuration
file, they are not being propagated when using
impersonation.
We need to propagate this configuration down to
the impersonated
process.

Closes #2920 from edgarRd/erod-propagate-hive-conf

NielsZeilemaker committed:
[AIRFLOW-1982] Fix Executor event log formatting

Closes #2925 from
NielsZeilemaker/fix_email_templating

NielsZeilemaker committed:
[AIRFLOW-1976] Fix for missing log/logger attribute FileProcessHandler

Closes #2922 from
NielsZeilemaker/fix_log_file_process_handler

ms32035 committed:
[AIRFLOW-1958] Add **kwargs to send_email

Closes #2908 from ms32035/email_kwargs

pdambrauskas committed:
[AIRFLOW-1480] Render template attributes for ExternalTaskSensor fields

Closes #2926 from pdambrauskas/feature/external_ta
sk_sensor_templates

bcb committed:
[AIRFLOW-1975] Make TriggerDagRunOperator callback optional

Closes #2921 from bcb/make-trigger-dag-run-
callback-optional

Commit:
Merge remote-tracking branch 'apache/master'

Commit:
[AIRFLOW-1688] Fix flake8 formatting issues

Commit:
[AIRFLOW-1688] Fix flake8 formatting issues

Commit:
[AIRFLOW-1688] Merge branch 'master' of https://github.com/albertocalderari/incubator-airflow

Commit:
[AIRFLOW-1688] Solved flake 8 issues in the test file as well
